src/ArrayCache.php

```diff
@@ -47,19 +47,42 @@ public function get($key)
             return Promise\resolve();
         }
 
+        if ($this->data[$key]['expires'] === null) {
+            // remove and append to end of array to keep track of LRU info
+            $item = $this->data[$key];
+            unset($this->data[$key]);
+            $this->data[$key] = $item;
+            return Promise\resolve($this->data[$key]['value']);
+        }
+
+        if ($this->data[$key]['expires'] < time()) {
+            unset($this->data[$key]);
+            return Promise\resolve();
+        }
+
+
         // remove and append to end of array to keep track of LRU info
         $value = $this->data[$key];
         unset($this->data[$key]);
         $this->data[$key] = $value;
-
-        return Promise\resolve($value);
+        return Promise\resolve($this->data[$key]['value']);
     }
 
-    public function set($key, $value)
+    public function set($key, $value, $ttl = null)
     {
+        $expires = null;
+
+        if (is_int($ttl)) {
+            $expires = time() + $ttl;
+        }
+
         // unset before setting to ensure this entry will be added to end of array
         unset($this->data[$key]);
-        $this->data[$key] = $value;
+        $this->data[$key] = [
+            'value' => $value,
+            'expires' => $expires,
+        ];
```

Comment:
Not a big fan of mixing these data structures. Probably makes more sense to store expiration in a separate array or possibly SplPriorityQueue?

Reply:
@clue good point, changed it :+1: